### PR TITLE
Initial release of golang Skydive operator

### PR DESCRIPTION
**Pull request by oranichu**
This is a work in progress for the skydive-golang operator.
Right now the operator can create all the deployment services and routes needed for the analyzer and also the replica set for the agents.
Feel free to comment, correct and add whatever you think is releveant

### CODE REVIEW

**Comment by amorenoz** [comment]
cc @russellb

File: README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+3. Clone this repo into your $GOPATH/src folder
+4. From skydive-operator folder run:
+
+```bash
+make manifests
+kubectl create -f config/crd/bases
+kubectl create -f config/samples/skydive_v1beta1_skydiveanalyzer.yaml 
+kubectl create -f config/samples/skydive_v1beta1_skydiveagents.yaml
```
Comment (eranra) [inline README.md]: (1) change the name from skydive_v1beta1_skydiveanalyzer to skydive_v1beta1_skydive_analyzer and from skydive_v1beta1_skydiveagents to skydive_v1beta1_skydive_agents
(2) move outside of the samples folder

File: README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+3. Clone this repo into your $GOPATH/src folder
+4. From skydive-operator folder run:
+
+```bash
+make manifests
+kubectl create -f config/crd/bases
+kubectl create -f config/samples/skydive_v1beta1_skydiveanalyzer.yaml 
+kubectl create -f config/samples/skydive_v1beta1_skydiveagents.yaml 
+make run
```
Comment (eranra) [inline README.md]: Add infomration on printing more and less debug infomration

**Comment by russellb** [comment]
Thanks for working on this!  One quick thing I noticed is that for this upstream operator, we can't do anything that is OpenShift specific, which includes the `Route` resource.  Maybe it'll be ok to include some optional OpenShift features that get used when the operator detects it's on OpenShift, but it'll need to work on non-OpenShift kubernetes clusters, as well.  I'd try it out on a `kind` cluster to validate that.

File: controllers/skydiveagents_controller.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// +kubebuilder:rbac:groups=skydive.example.com,resources=skydiveagents,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=skydive.example.com,resources=skydiveagents/status,verbs=get;update;patch
+func (r *SkydiveAgentsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("skydiveanalyzer", req.NamespacedName)
+
+	log.Info("initializing assets for skydive-agents")
+	assets := manifests.NewAssets("assets")
+
+	log.Info("Building configuration")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
```
Comment (eranra) [inline controllers/skydiveagents_controller.go]: Replace with

```	
var config *rest.Config
var err error
if kubeConfig := os.Getenv("KUBECONFIG"); kubeConfig != "" {
	config, err = clientcmd.BuildConfigFromFlags("", kubeConfig)
} else {
	config, err = rest.InClusterConfig()
}
```

File: README.md (hunk 1 above)
Comment (oranichu) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (oranichu) [reply README.md]: What did you mean here?

File: controllers/skydiveagents_controller.go (hunk 3 above)
Comment (oranichu) [reply controllers/skydiveagents_controller.go]: So in my case, I'll have to add an ENV var named KUBECONFIG and set its path right?

Comment (eranra) [reply controllers/skydiveagents_controller.go]: @oranichu yes or if you are logged into a cluster using kubectl in the current shell you should also be covered

File: README.md (hunk 2 above)
Comment (eranra) [reply README.md]: @oranichu as we spoke on the phone ... manage the level of logs printed when executing `make run`

File: README.md (hunk 1 above)
Comment (rcarrillocruz) [reply README.md]: There is a mix of oc and kubectl commands.
My understanding is that we are trying to make this upstream first, thus I'd change the commands to just kubectl.

If there are specific options to make this to work on OpenShift, then it's worth having a README_OPENSHIFT.md

Comment (oranichu) [reply README.md]: Pushed a script that deploys it on openshift
still working on making it kubectl friendly

File: README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+<!-- ABOUT THE PROJECT -->
+
+# Skydive Operator (WIP)
```
Comment (eranra) [inline README.md]: remove (WIP)

File: README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+This is a simple operator to deploy skydive analyzer and agents.
```
Comment (eranra) [inline README.md]: remove the word simple, add flow exporter

File: README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+### Prerequisites
+
+* Make sure you have golang installed on your machine, The go-lang version that this operator was built on is go1.15.7
+  darwin/amd64
+
+* An openshift cluster
+    * If you wish to run this locally by using kind please
+      check [KindInstallationREADME.md](hack/KindInstallationREADME.md)
+
+### Modifying the CRD
```
Comment (eranra) [inline README.md]: move after installation section (and you can call this section Customize skydive deployment / CRD)

File: README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+Change Env var : (TODO: more info to be added here)
+
+Choose your logging level (defaults to DEBUG)
+
+### Installation - Open-Shift
+
+run the script :
+
+```
+bash deploy_skydive_operator_on_openshift.sh
```
Comment (eranra) [inline README.md]: remvoe the word bash and add ./

Comment (eranra) [reply README.md]: All add the text ... make sure that you are logged into open shift cluster

File: deploy_skydive_operator_on_openshift.sh [hunk 8]
```diff
@@ -0,0 +1,52 @@
+# delete skydive project if it exists (to get new frech deployment)
```
Comment (eranra) [inline deploy_skydive_operator_on_openshift.sh]: s/frech/fresh

File: deploy_skydive_operator_on_openshift.sh [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+oc_create_skydive_project() {
+  oc adm new-project --node-selector='' skydive
+  oc project skydive
+}
+
+# set credentials for skydive deployment
+oc_set_credentials() {
+  # analyzer and agent run as privileged container
+  oc adm policy add-scc-to-user privileged -z default
+  # analyzer need cluster-reader access get all informations from the cluster
```
Comment (eranra) [inline deploy_skydive_operator_on_openshift.sh]: s/informations/information

File: deploy_skydive_operator_on_openshift.sh [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+oc_clear_skydive_crd() {
+SKYDIVE_SUITE_CRD=$(oc get skydive)
+if [ ! -z "$SKYDIVE_SUITE_CRD" ]; then
+  oc delete -f config/crd/bases/skydive.example.com_skydives.yaml
+fi
+}
+
+# deoploy skydive
```
Comment (eranra) [inline deploy_skydive_operator_on_openshift.sh]: s/deoploy/deploy

File: deploy_skydive_operator_on_openshift.sh (hunk 9 above)
Comment (eranra) [inline deploy_skydive_operator_on_openshift.sh]: s/get/to get

File: deploy_skydive_operator_on_openshift.sh [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+oc_set_credentials() {
+  # analyzer and agent run as privileged container
+  oc adm policy add-scc-to-user privileged -z default
+  # analyzer need cluster-reader access get all informations from the cluster
+  oc adm policy add-cluster-role-to-user cluster-reader -z default
+}
+
+oc_clear_skydive_crd() {
+SKYDIVE_SUITE_CRD=$(oc get skydive)
+if [ ! -z "$SKYDIVE_SUITE_CRD" ]; then
```
Comment (eranra) [inline deploy_skydive_operator_on_openshift.sh]: s/! -z/-n

**Comment by eranra** [comment]
@oranichu I started the operator today and got::

```2021-03-14T16:12:40.456+0200    ERROR   setup   problem running manager {"error": "no matches for kind \"SkydiveFlowExporter\" in version \"skydive.example.com/v1\""}
```

You need to deploy the CRD for FlowExporter in the skydive deployment script (and vice versa)

After running 
`kubectl create -f config/crd/bases/skydive.example.com_skydiveflowexporters.yaml`

It works :-)

**Comment by rcarrillocruz** [comment]
Apologies for the delay, I was meant to suggest how to put the OpenShift certs for OVN visualization:

https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/ovnkube-node.yaml#L328

There you can see the configmap and secrets used, in case you want to put that on the openshift script.

**Comment by eranra** [comment]
@oranichu  ... so .... you will need to download those / copy the infomration from the `openshift-ovn-kubernetes` namespace to the skydive deployment namespace (configmaps can't be shared between namespaces)  ... than use them in skydive operator here: https://github.com/skydive-project/skydive/blob/master/etc/skydive.yml.default#L178

@rcarrillocruz above correct ??? 

image

**Comment by rcarrillocruz** [comment]
That's right, since this is in out-of-band script what you could do is oc -n openshift-ovn-kubernetes get <cm|secret> <blah> -o yaml > ovn-<cm|secret>.yaml, then pipe it to create it on the skydive operator namespace, that could be a way of doing it.

**Comment by eranra** [comment]
Need to add this to the script ::
-=--==-=-=-=-

`oc get configmap ovn-ca --namespace=openshift-ovn-kubernetes -o yaml | sed 's/namespace: openshift-ovn-kubernetes/namespace: skydive/' | oc  apply --namespace=skydive -f -`

`oc get secret ovn-cert --namespace=openshift-ovn-kubernetes -o yaml | sed 's/namespace: openshift-ovn-kubernetes/namespace: skydive/' | oc apply --namespace=skydive -f -`

`MASTER_IP=$(kubectl get node --selector='!node-role.kubernetes.io/master' -o jsonpath='{.items[0].status.addresses[?(@.type=="InternalIP")].address}')`

In my case `MASTER_IP` was 10.0.139.85

And I changed analyzer deployment yaml to include::

```
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_ADDRESS
              value: tcp:10.0.139.85:9641
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_CA
              value: /ovn-cert/tls.crt
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_KEY
              value: /ovn-cert/tls.key
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_CACERT
              value: /ovn-ca/ca-bundle.crt

          volumeMounts:
          - mountPath: /ovn-cert
            name: ovn-cert
          - mountPath: /ovn-ca
            name: ovn-ca

      volumes:
      - name: ovn-ca
        configMap:
          name: ovn-ca
      - name: ovn-cert
        secret:
          secretName: ovn-cert
          
 ```       

I still get this error in skydive analyzer log::


`skydive-analyzer-6f97576954-rwhs2: Failed to start: failed to connect: dial tcp 10.0.139.85:9641: connect: connection refused, retrying in background every 3s`


I changed manually the IP to `10.0.175.118` which is the POD IP (need to understand what IP we need to use, it looks like OVN is not exposing the NB so we can't hook to Node IP :-) 

Now the code connects but there is probably something wrong with OVN ... I see in skydive logs ::


```
2021-03-21T16:33:21.871Z	[34mINFO[0m	http/server.go:105 (*Server).Listen	skydive-analyzer-6c969c6d6f-bllpr: Listening on socket 0.0.0.0:8082
2021-03-21T16:33:21.885Z	[34mINFO[0m	client/election.go:104 (*MasterElector).start	skydive-analyzer-6c969c6d6f-bllpr: starting as the master for /elections/alert-server: skydive-analyzer-6c969c6d6f-bllpr
2021-03-21T16:33:21.897Z	[34mINFO[0m	k8s/cluster.go:54 (*clusterCache).addClusterNode	skydive-analyzer-6c969c6d6f-bllpr: Added cluster{Name: cluster}
2021-03-21 16:33:21.902222 I | ListDbs failureunexpected EOF
```



I also tried from with in the container to execute::

```curl -k --cert /ovn-cert/tls.crt --key /ovn-cert/tls.key --cacert /ovn-ca/ca-bundle.crt https://10.0.175.118:9641/```

but I get empty response from server :-(

**Comment by eranra** [comment]
@rcarrillocruz can you look on the above and advice ... maybe I did something wrong in the configuration ... mixed something ??? of maybe there is something that needs more work in OVN code ???

@oranichu @russellb  ^^^ FYI

**Comment by eranra** [comment]
@oranichu BTW this discussion with @rcarrillocruz and the work to enable OVN support should not block this PR from merge ... we should merge this PR and add OVN support as next phase of functionality against the operator.

@lebauce FYI, hope you agree.

**Comment by rcarrillocruz** [comment]
We do expose the OVN NB with a service, but it's a headless service:

https://github.com/openshift/cluster-network-operator/blob/dfea1dc53ed82f5e4a4e85b7e8e863af3e4bd54a/bindata/network/ovn-kubernetes/005-service.yaml

This is how the CNO finds out the IP, by getting the svc endpoints:

https://github.com/openshift/cluster-network-operator/blob/dfea1dc53ed82f5e4a4e85b7e8e863af3e4bd54a/bindata/network/ovn-kubernetes/005-service.yaml

So the script should be doing something similar.

I'm swamped the next couple of weeks with end of release stuff, will try to find some time to debug it but cannot give you timelines.

**Comment by eranra** [comment]
@rcarrillocruz thanks ... we will need to look on that later ... NP

File: hack/KindInstallationREADME.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  
+
+### Installation
+
+#### Kind
+Install kind on your machine, use [this](https://kind.sigs.k8s.io/docs/user/quick-start/) quick start guide for help, 
+make sure that your have a running kind cluster and that kubectl commands refer to relevant context.
+
+After installing and running the Kind cluster, make sure the env var KUBECONFIG is set to the correct kubeConfig file,
+it usually defaults to $HOME/.kube/config, run this command:
```
Comment (jotak) [inline hack/KindInstallationREADME.md]: I think this part is also relevant for the openshift install, not only Kind. I just run into the issue, with this error in logs : `unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined`. I had no KUBECONFIG set up (I'm connecting to a remote CRC with `oc`. It works fine but doesn't set up the relevant kubeconfig when I log in, so I had to pull it manually from my remote crc)

**Pull request merged**